File: pytools/examples/php/server/protocol/MsgTestXX.php
<?php
namespace protocol;

use network\Packet;


class MsgTestXX
{
	private $idU8 = 0;
	private $idF32 = [];
	private $idOpU8Flag = 0;
	private $idOpU8 = 0;


	public function __construct(&$packet=null)
	{
		if ($packet)
		{
			$this->decode($packet);
		}
	}

	public function decode(&$packet)
	{
		$this->idU8 = $packet->readU8();
		$idF32Count = $packet->readU16();
		for ($i = 0; $i < $idF32Count; $i++)
		{
			array_push($this->idF32, $packet->readF32());
		}
		$this->idOpU8Flag = $packet->readU8();
		if ($this->idOpU8Flag == 1)
		{
			$this->idOpU8 = $packet->readU8();
		}
	}

	public function encode()
	{
		$packet = new Packet();
		$packet->writeU8($this->idU8);
		$idF32Count = count($this->idF32);
		$packet->writeU16($idF32Count);
		foreach ($this->idF32 as $idF32)
		{
			$packet->writeF32($idF32);
		}
		$packet->writeU8($this->idOpU8Flag);
		if ($this->idOpU8Flag == 1)
		{
			$packet->writeU8($this->idOpU8);
		}
		return $packet->readBytes();
	}

	public function getBytes()
	{
		return $this->encode();
	}


	public function setIdU8($idU8)
	{
		$this->idU8 = $idU8;
	}
	public function getIdU8()
	{
		return $this->idU8;
	}

	public function setIdF32($idF32)
	{
		$this->idF32 = $idF32;
	}
	public function getIdF32()
	{
		return $this->idF32;
	}

	public function setIdOpU8($idOpU8)
	{
		$this->idOpU8Flag = 1;
		$this->idOpU8 = $idOpU8;
	}
	public function getIdOpU8()
	{
		return $this->idOpU8;
	}

}

File: pytools/data/data.code.xlua.client/MsgTestXX.lua
MsgTestXX = {}
setmetatable(MsgTestXX, {__index = _G})
local _ENV = MsgTestXX


function new()
	local tab_obj = {}
	setmetatable(tab_obj, {__index = MsgTestXX})
	tab_obj:init()
	return tab_obj
end

function init(self)
	self.id_u8 = 0
	self.id_f32 = {}
	self.id_op_u8_flag = 0
	self.id_op_u8 = 0
end


function encode(self)
	pack = packet.new()
	pack:writeU8(self.id_u8)
	id_f32_count = #self.id_f32
	pack:writeU16(id_f32_count)
	for i = 1, id_f32_count do
		pack:writeF32(table.remove(self.id_f32))
	end
	pack:writeU8(self.id_op_u8_flag)
	if self.id_op_u8_flag then
		pack:writeU8(self.id_op_u8)
	end
	return pack:readBytes()
end

function decode(self, pack)
	self.id_u8 = pack:readU8()
	id_f32_count = pack:readU16()
	for i = 1, id_f32_count do
		table.insert(self.id_f32, pack:readF32())
	end
	self.id_op_u8_flag = pack:readU8()
	if self.id_op_u8_flag then
		self.id_op_u8 = pack:readU8()
	end
	return self
end

function getBytes(self)
	return self:encode()
end


function setIdU8(self, id_u8)
	self.id_u8 = id_u8
end
function getIdU8(self)
	return self.id_u8
end

function setIdF32(self, id_f32)
	self.id_f32 = id_f32
end
function getIdF32(self)
	return self.id_f32
end

function setIdOpU8(self, id_op_u8)
	self.id_op_u8_flag = 1
	self.id_op_u8 = id_op_u8
end
function getIdOpU8(self)
	return self.id_op_u8
end

